src/ipred_prepare.rs

```diff
@@ -182,177 +193,131 @@ pub unsafe fn rav1d_prepare_intra_edges<BD: BitDepth>(
             || av1_intra_prediction_edges[mode as usize]
                 .needs
                 .contains(Needs::LEFT)
-                && have_left == 0)
+                && !have_left)
     {
-        if !prefilter_toplevel_sb_edge.is_null() {
-            dst_top = &*prefilter_toplevel_sb_edge.offset((x * 4) as isize) as *const BD::Pixel;
+        let px_have = cmp::min(8 * tw, 4 * (w - x)) as usize;
+        let n = px_have + have_left as usize;
+        if let Some(prefilter_toplevel_sb_edge) = prefilter_toplevel_sb_edge {
+            let offset = (x * 4) as usize - have_left as usize;
+            &prefilter_toplevel_sb_edge[offset..offset + n]
```

Comment:
```suggestion
            &prefilter_toplevel_sb_edge[offset..][..n]
```

Reply:
> Just a neater, more readable pattern I think.

Actually, I found this is hard for compiler to get [optimized](https://godbolt.org/z/Mzd9a6E3b), maybe we should keep it.

Reply:
Hi @CrazyboyQCD!  Thanks for pointing this out!  I wasn't aware this optimized poorly sometimes, as there are other cases I've checked where `[offset..][..len]` was equivalent or better than `[offset..offset + len]`, though I don't remember the specifics.

I'm surprised rustc/LLVM is worse at optimizing `[offset..][..len]`, which I narrowed down a bit more to these two nearly identical versions:

https://godbolt.org/z/n76YcdEP5

```rust
pub fn f1(a: &[u8], b: usize, c: usize) -> &[u8] {
    if a.len() > b + c && a.len() > b && b + c > b {
        &a[b..b + c]
        // &a[b..][..c]
    } else {
        &[]
    }
}

pub fn f2(a: &[u8], b: usize, c: usize) -> &[u8] {
    if a.len() > b + c && a.len() > b && b + c > b {
        &a[b..b + c]
    } else {
        &[]
    }
}
```

Have you considered opening this as a rustc bug, as I don't see why one version should be optimized much worse?

Also, although in this case `[offset..offset + len]` optimized better, that's not always the case, and we can't generalize.  For example, for b861557, `[offset..][..len]` is optimized better instead:

https://godbolt.org/z/MjPEjbnsW

```rust
pub fn corner1(topleft_out: &mut [u8], topleft_out_offset: usize) -> &mut [u8; 3] {
    <&mut [_; 3]>::try_from(&mut topleft_out[topleft_out_offset - 1..=topleft_out_offset + 1]).unwrap()
}

pub fn corner2(topleft_out: &mut [u8], topleft_out_offset: usize) -> &[u8] {
    <&mut [_; 3]>::try_from(&mut topleft_out[topleft_out_offset - 1..][..3]).unwrap()
}
```

`corner1` contains an extra check and call to `slice_end_index_overflow_fail` that's not there for `corner2`.  Though this might just be due to `..=` vs. `..`, as `..=` usually seems to optimize worse.  So for this case, @fbossen, could you rebase and drop b861557?

Either way, I don't think we should generalize this difference in optimizations.  It may well change in future versions of rustc and LLVM.

Reply:
@kkysen I'm glad to see further exploration about this.
> https://godbolt.org/z/s5bEzhaTa
> ```rust
> pub fn corner1(topleft_out: &mut [u8], topleft_out_offset: usize) -> &mut [u8; 3] {
>     <&mut [_; 3]>::try_from(&mut topleft_out[topleft_out_offset - 1..topleft_out_offset + 2]).unwrap()
> }
>
> pub fn corner2(topleft_out: &mut [u8], topleft_out_offset: usize) -> &mut [u8; 3] {
>     <&mut [_; 3]>::try_from(&mut topleft_out[topleft_out_offset - 1..][..3]).unwrap()
> }
> ```
In the link, I didn't see the output you have, what I saw is that they both have two calls to `index_overflow_fail`, my output is on [the clipboard](https://goonlinetools.com/online-clipboard/#3920). And after I added same [asserts](https://godbolt.org/z/nEv939zxs) in them, the `corner1` seems to be better optimized(I use `unreachable_unchecked` to get cleaner output).
Anyway, I agree that this shoud be reported to rust and we could keep the changes.

Reply:
> In the link, I didn't see the output you have, what I saw is that they both have two calls to `index_overflow_fail`, my output is on [the clipboard](https://goonlinetools.com/online-clipboard/#3920).

Whoops, I deleted the `..=` that was in b861557, which made them the same again.

Reply:
Alright, I see the diffrence now.

Reply:
I also found that if you re-arrange the precondition checks/assumes to more closely follow the `[offset..][..len]` version, you can get even more optimized code:

https://godbolt.org/z/ve83f8Gjb

```rust
pub fn f3(a: &[u8], b: usize, c: usize) -> &[u8] {
    if a.len() > b {
        let a = &a[b..];
        if a.len() > c {
            return &a[..c];
        }
    }
    &[]
}
```